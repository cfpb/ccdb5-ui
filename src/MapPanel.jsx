import './MapPanel.less'
import ActionBar from './ActionBar'
import { connect } from 'react-redux'
import DateIntervals from './DateIntervals'
import Loading from './Dialogs/Loading'
import React from 'react'
import RowChart from './RowChart'
import TileChartMap from './TileChartMap'

export class MapPanel extends React.Component {
  render() {
    return (
      <section className="map-panel">
        <ActionBar/>
        <DateIntervals/>
        <TileChartMap/>
        <RowChart aggtype="product" />
        <RowChart aggtype="issue" />
        <Loading isLoading={ this.props.isLoading || false }/>
      </section>
    )
  }
}

const mapStateToProps = state => ( {
  isLoading: state.results.isLoading,
  items: state.results.items
} )

export default connect( mapStateToProps )( MapPanel )
