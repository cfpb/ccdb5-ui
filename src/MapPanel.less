@import (less) "base.less";

.map-panel {
  .cards-panel {
    padding: @grid_gutter-width;
    margin: 0;
 
   @media @phone {
    padding: @gutter-normal;
  }
 }

  h2 {
    padding: @gutter-normal;    
  }
}