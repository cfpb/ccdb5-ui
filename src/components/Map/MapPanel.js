import '../RefineBar/RefineBar.less';
import { ActionBar } from '../ActionBar/ActionBar';
import { useDispatch, useSelector } from 'react-redux';
import ErrorBlock from '../Warnings/Error';
import { FilterPanel } from '../Filters/FilterPanel';
import { FilterPanelToggle } from '../Filters/FilterPanelToggle';
import { Loading } from '../Loading/Loading';
import { MapToolbar } from './MapToolbar';
import { mapWarningDismissed } from '../../reducers/filters/filtersSlice';
import { PerCapita } from '../RefineBar/PerCapita';
import { processRows } from '../../utils/chart';

import { useMemo } from 'react';
import { RowChart } from '../Charts/RowChart/RowChart';
import { Separator } from '../RefineBar/Separator';
import { TabbedNavigation } from '../TabbedNavigation';
import { TileChartMap } from './TileChartMap/TileChartMap';
import Warning from '../Warnings/Warning';
import { selectAggsTotal } from '../../reducers/aggs/selectors';

import {
  selectFiltersEnablePer1000,
  selectFiltersMapWarningEnabled,
} from '../../reducers/filters/selectors';
import {
  selectMapActiveCall,
  selectMapError,
  selectMapResults,
} from '../../reducers/map/selectors';

import {
  selectQueryDateReceivedMax,
  selectQueryDateReceivedMin,
} from '../../reducers/query/selectors';

import {
  selectViewExpandedRows,
  selectViewWidth,
} from '../../reducers/view/selectors';

import { shortFormat } from '../../utils';

const WARNING_MESSAGE =
  '“Complaints per 1,000 population” is not available with your filter ' +
  'selections.';

const MAP_ROWCHART_HELPERTEXT =
  'Product the consumer identified in the complaint. Click on a product ' +
  'to expand sub-products';

export const MapPanel = () => {
  const dispatch = useDispatch();
  const total = useSelector(selectAggsTotal);

  const enablePer1000 = useSelector(selectFiltersEnablePer1000);
  const mapWarningEnabled = useSelector(selectFiltersMapWarningEnabled);

  const activeCall = useSelector(selectMapActiveCall);
  const results = useSelector(selectMapResults);
  const hasError = useSelector(selectMapError);

  const maxDate = useSelector(selectQueryDateReceivedMax);
  const minDate = useSelector(selectQueryDateReceivedMin);

  const expandedRows = useSelector(selectViewExpandedRows);
  const width = useSelector(selectViewWidth);
  const hasMobileFilters = width < 750;
  const hasWarning = !enablePer1000 && mapWarningEnabled;
  const productData = useMemo(() => {
    return processRows(results.product, false, 'Product', expandedRows);
  }, [results, expandedRows]);

  const MAP_ROWCHART_TITLE = `Product by highest complaint volume ${shortFormat(
    minDate,
  )} to ${shortFormat(maxDate)}`;

  const onDismissWarning = () => {
    dispatch(mapWarningDismissed());
  };

  return (
    <section className="map-panel">
      <ActionBar />
      <TabbedNavigation />
      {!!hasError && (
        <ErrorBlock text="There was a problem executing your search" />
      )}
      {!!hasWarning && (
        <Warning text={WARNING_MESSAGE} closeFn={onDismissWarning} />
      )}
      {!!hasMobileFilters && <FilterPanel />}
      <FilterPanelToggle />
      <div className="layout-row refine-bar">
        <Separator />
        <PerCapita />
      </div>
      <TileChartMap />
      <MapToolbar />
      <RowChart
        id="product"
        colorScheme={productData.colorScheme}
        data={productData.data}
        title={MAP_ROWCHART_TITLE}
        helperText={MAP_ROWCHART_HELPERTEXT}
        total={total}
      />

      <Loading isLoading={activeCall !== ''} />
    </section>
  );
};
