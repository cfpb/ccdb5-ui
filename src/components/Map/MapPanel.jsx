import '../RefineBar/RefineBar.less'
import ActionBar from '../ActionBar'
import { connect } from 'react-redux'
import DateRanges from '../RefineBar/DateRanges'
import ErrorBlock from '../Warnings/Error'
import FilterPanel from '../Filters/FilterPanel'
import FilterPanelToggle from '../Filters/FilterPanelToggle'
import Loading from '../Dialogs/Loading'
import MapToolbar from './MapToolbar'
import { mapWarningDismissed } from '../../actions/view'
import { PerCapita } from '../RefineBar/PerCapita'
import { processRows } from '../../utils/chart'
import React from 'react'
import RowChart from '../Charts/RowChart'
import { Separator } from '../RefineBar/Separator'
import { shortFormat } from '../../utils'
import TabbedNavigation from '../TabbedNavigation'
import TileChartMap from '../Charts/TileChartMap'
import Warning from '../Warnings/Warning'


const WARNING_MESSAGE = '“Complaints per' +
  ' 1,000 population” is not available with your filter selections.'

const MAP_ROWCHART_HELPERTEXT = 'Product the consumer identified in the' +
' complaint. Click on a product to expand sub-products'

export class MapPanel extends React.Component {
  // eslint-disable-next-line complexity
  render() {
    const MAP_ROWCHART_TITLE = 'Product by highest complaint volume' +
      ' ' + this.props.minDate + ' to ' + this.props.maxDate

    return (
      <section className="map-panel">
        <ActionBar/>
        <TabbedNavigation />
        { this.props.error &&
          <ErrorBlock text="There was a problem executing your search" />
        }
        { this.props.showWarning &&
          <Warning text={ WARNING_MESSAGE }
                   closeFn={this.props.onDismissWarning}/> }
        { this.props.showMobileFilters && <FilterPanel/> }
        <div className="layout-row refine-bar">
          <FilterPanelToggle/>
          <Separator />
          <DateRanges/>
          <PerCapita/>
        </div>
        <TileChartMap/>
        <MapToolbar/>
        <RowChart id="product"
                  colorScheme={this.props.productData.colorScheme}
                  data={this.props.productData.data}
                  title={ MAP_ROWCHART_TITLE }
                  helperText={ MAP_ROWCHART_HELPERTEXT }
                  total={ this.props.total }/>

        <Loading isLoading={ this.props.isLoading || false }/>
      </section>
    )
  }
}

const mapStateToProps = state => {
  const {
    error,
    isLoading,
    results
  } = state.map

  const {
    date_received_max: maxDate,
    date_received_min: minDate,
    enablePer1000,
    mapWarningEnabled
  } = state.query

  const { expandedRows, width } = state.view

  return {
    error,
    isLoading,
    minDate: shortFormat( minDate ),
    maxDate: shortFormat( maxDate ),
    productData: processRows( results.product, false, 'Product', expandedRows ),
    showMobileFilters: width < 750,
    showWarning: !enablePer1000 && mapWarningEnabled,
    total: state.aggs.total
  }
}

export const mapDispatchToProps = dispatch => ( {
  onDismissWarning: () => {
    dispatch( mapWarningDismissed() )
  }
} )


export default connect( mapStateToProps, mapDispatchToProps )( MapPanel )
