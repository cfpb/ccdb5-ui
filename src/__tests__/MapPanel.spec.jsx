import configureMockStore from 'redux-mock-store'
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux'
import MapPanel from '../MapPanel';
import React from 'react'
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk'

const fixture = [
  { key: 'CA', doc_count: 62519 },
  { key: 'FL', doc_count: 47358 }
]

function setupSnapshot(items=[], initialStore={}) {
  const results = Object.assign({
    doc_count: 100,
    error: '',
    hasDataIssue: false,
    isDataStale: false,
    items,
    total: items.length
  }, initialStore);

  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)
  const store = mockStore({
    map: {
      state: []
    },
    query: {
      from: 0,
      size: 10,
      tab: 'Map'
    },
    results
  })

  return renderer.create(
    <Provider store={ store } >
      <IntlProvider locale="en">
        <MapPanel items={ items } />
      </IntlProvider>
    </Provider>
  )
}

describe('component:MapPanel', () => {
  it('renders without crashing', () => {
    const target = setupSnapshot(fixture)
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });
})
