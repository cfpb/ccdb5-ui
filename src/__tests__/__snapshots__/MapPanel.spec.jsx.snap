// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:MapPanel renders without crashing 1`] = `
<section
  className="map-panel"
>
  <summary
    className="action-bar"
  >
    <div>
      <h2>
        Showing 
        <span>
          1
        </span>
         matches out of 
        <span>
          100
        </span>
         total complaints
      </h2>
    </div>
    <div
      className="layout-row"
    >
      <h4
        className="flex-all"
      >
        <button
          className="a-btn a-btn__link"
          data-gtm_ignore="true"
          onClick={[Function]}
        >
          Export results
        </button>
      </h4>
    </div>
  </summary>
  <div />
  <section
    className="date-intervals"
  >
    <button
      className="date-selector "
      onClick={[Function]}
    >
      3m
    </button>
    <button
      className="date-selector "
      onClick={[Function]}
    >
      6m
    </button>
    <button
      className="date-selector "
      onClick={[Function]}
    >
      1y
    </button>
    <button
      className="date-selector "
      onClick={[Function]}
    >
      3y
    </button>
    <button
      className="date-selector "
      onClick={[Function]}
    >
      All
    </button>
  </section>
  <div>
    <div
      className="cfpb-chart"
      data-chart-type="tile_map"
      id="tile-chart-map"
    />
  </div>
</section>
`;
